[pytorch] change codegen dispatch key from string to enum

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51115 [pytorch] change codegen dispatch key from string to enum**

Summary:
Add enum type for dispatch key. Prepare to implement the DispatchTable
computation logic in python for static dispatch.

Verified byte-for-byte compatibility of the codegen output.

Differential Revision: [D26077430](https://our.internmc.facebook.com/intern/diff/D26077430)

ljk53 committed:
[pytorch] change codegen dispatch key from string to enum

Summary:
Add enum type for dispatch key. Prepare to implement the DispatchTable
computation logic in python for static dispatch.

Verified byte-for-byte compatibility of the codegen output.

[ghstack-poisoned]

ailzhang commented on tools/codegen/model.py:
```diff
@@ -47,11 +47,89 @@ def __str__(self) -> str:
 # Valid values of the 'variants' field in native_functions.yaml
 Variant = Enum('Variant', ('function', 'method'))
 
+# NOTE: Keep the list in sync with `DispatchKey` in c10/core/DispatchKey.h
```
I wonder if it's possible to add a check to enforce the parity between the two lists? Otherwise the registration can easily go to the wrong backends. :P

bhosmer replied:
@ailzhang I think we're actually safe from that, since we're not generating the enum's underlying values into code - it's still the symbolic names via `__str__` (hence byte-for-byte compatible output).

That said, we could think about adding parity enforcement later on - say, a static assert that verified the same total number of distinct dispatch keys - not sure the best generated code to add this to though.

Better yet would be a single source of truth. I think our two alternatives are
* actually codegen the C++ definition (I think we have a cutout that prevents regeneration of unchanged artifacts that we could use, foggy on details though)
* read the C++ definition into python, though doing this without actually parsing C++ would probably involve changes that would degrade UX on the C++ side (macros or whatever)

Anyway though, given that this PR makes things strictly better than before (AFAICT codegen used to just silently ignore invalid dispatch key names in `dispatch:` entries), I don't think any of the above is worth piling onto this PR unless it's dead easy/obvious.

bhosmer replied:
> actually codegen the C++ definition 

...this fits in with the broader idea of lifting the entire dispatch key data model into metadata and using it to generate not just this enum but also helper functions and so on

ljk53 replied:
Thanks @bhosmer for the rely - I was about to comment the same :p

I was thinking of adding higher level parity check in subsequent PRs where I do dispatch table computation.
I also thought about codegen / macro + parsing / macro + test-method-to-print-all. Might do one of these later :)

bhosmer commented on tools/codegen/gen.py:
```diff
@@ -1283,24 +1290,24 @@ def make_file_manager(install_dir: str) -> FileManager:
     # NB: substrings in these dispatch keys matter, we do tests to see if
```
No need for this comment anymore 🎉

bhosmer commented:
@ljk53 didn't run down the build errors here and on phab though

ljk53 committed:
Update on "[pytorch] change codegen dispatch key from string to enum"


Summary:
Add enum type for dispatch key. Prepare to implement the DispatchTable
computation logic in python for static dispatch.

Verified byte-for-byte compatibility of the codegen output.

Differential Revision: [D26077430](https://our.internmc.facebook.com/intern/diff/D26077430)

[ghstack-poisoned]

ljk53 commented:
> @ljk53 didn't run down the build errors here and on phab though

Fixed. It's the backend blacklist - it was looking up strings in the enum set.
